# Joining Data with Pandas

***

## Data Merging Basics

### Importing Libraries

In [1]:
# Import libarary
import pickle
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np

In [2]:
# Load 1st dataset tax_owners

taxi_owners = pickle.load(
    open("D:/git_repositories/Datacamp-Joining_data_with_pandas/Datasets/taxi_owners.p", "rb")
)

In [3]:
# Load 2nd dataset taxi_vehicles

taxi_veh = pickle.load(
    open("D:/git_repositories/Datacamp-Joining_data_with_pandas/Datasets/taxi_veh.p", "rb")
)

---

### Showing Datasets

In [6]:
# 1st data
taxi_owners.head()

rid   vid           owner                 address    zip
0  T6285  6285  AGEAN TAXI LLC     4536 N. ELSTON AVE.  60630
1  T4862  4862    MANGIB CORP.  5717 N. WASHTENAW AVE.  60659
2  T1495  1495   FUNRIDE, INC.     3351 W. ADDISON ST.  60618
3  T4231  4231    ALQUSH CORP.   6611 N. CAMPBELL AVE.  60645
4  T5971  5971  EUNIFFORD INC.     3351 W. ADDISON ST.  60618

In [7]:
# 2nd data
taxi_veh.head()

vid    make   model  year fuel_type                owner
0  2767  TOYOTA   CAMRY  2013    HYBRID       SEYED M. BADRI
1  1411  TOYOTA    RAV4  2017    HYBRID          DESZY CORP.
2  6500  NISSAN  SENTRA  2019  GASOLINE       AGAPH CAB CORP
3  2746  TOYOTA   CAMRY  2013    HYBRID  MIDWEST CAB CO, INC
4  5922  TOYOTA   CAMRY  2013    HYBRID       SUMETTI CAB CO

#### Total Columns and Rows in these datasets

In [8]:
taxi_owners.shape

(3519, 5)

In taxi owners dataset there are 5 columns and 3,519 rows.

In [9]:
taxi_veh.shape

(3519, 6)

In taxi vehicles dataset there are 6 columns and 3,519 rows.

---

### Inner Join

**Merge taxi_owners with taxi_veh on the column vid, and save the result to taxi_own_veh**

In [4]:
# Merging two datasets

taxi_owners_veh = taxi_owners.merge(
    taxi_veh,       # merging 2nd dataset in 1st dataset
    on = 'vid'      # mergind id or same column 'vid'
)

In [12]:
# Printing out the new merged dataset
print(taxi_owners_veh.columns)

Index(['rid', 'vid', 'owner_x', 'address', 'zip', 'make', 'model', 'year',
       'fuel_type', 'owner_y'],
      dtype='object')


In [13]:
# Showing merged dataset
taxi_owners_veh.head()

rid   vid         owner_x                 address    zip    make   model  \
0  T6285  6285  AGEAN TAXI LLC     4536 N. ELSTON AVE.  60630  NISSAN  ALTIMA   
1  T4862  4862    MANGIB CORP.  5717 N. WASHTENAW AVE.  60659   HONDA     CRV   
2  T1495  1495   FUNRIDE, INC.     3351 W. ADDISON ST.  60618  TOYOTA  SIENNA   
3  T4231  4231    ALQUSH CORP.   6611 N. CAMPBELL AVE.  60645  TOYOTA   CAMRY   
4  T5971  5971  EUNIFFORD INC.     3351 W. ADDISON ST.  60618  TOYOTA  SIENNA   

   year fuel_type         owner_y  
0  2011    HYBRID  AGEAN TAXI LLC  
1  2014  GASOLINE    MANGIB CORP.  
2  2015  GASOLINE   FUNRIDE, INC.  
3  2014    HYBRID    ALQUSH CORP.  
4  2015  GASOLINE  EUNIFFORD INC.

**Set the left and right table suffixes for overlapping columns of the merge to _own and _veh, respectively.**

In [5]:
taxi_owners_veh = taxi_owners.merge(
    taxi_veh,
    on = 'vid',
    suffixes = ('_own', '_veh')
)

In [18]:
taxi_owners_veh.head()

rid   vid       owner_own                 address    zip    make   model  \
0  T6285  6285  AGEAN TAXI LLC     4536 N. ELSTON AVE.  60630  NISSAN  ALTIMA   
1  T4862  4862    MANGIB CORP.  5717 N. WASHTENAW AVE.  60659   HONDA     CRV   
2  T1495  1495   FUNRIDE, INC.     3351 W. ADDISON ST.  60618  TOYOTA  SIENNA   
3  T4231  4231    ALQUSH CORP.   6611 N. CAMPBELL AVE.  60645  TOYOTA   CAMRY   
4  T5971  5971  EUNIFFORD INC.     3351 W. ADDISON ST.  60618  TOYOTA  SIENNA   

   year fuel_type       owner_veh  
0  2011    HYBRID  AGEAN TAXI LLC  
1  2014  GASOLINE    MANGIB CORP.  
2  2015  GASOLINE   FUNRIDE, INC.  
3  2014    HYBRID    ALQUSH CORP.  
4  2015  GASOLINE  EUNIFFORD INC.

In [19]:
# select the fuel type column and counts the items
taxi_owners_veh['fuel_type'].value_counts()

HYBRID                    2792
GASOLINE                   611
FLEX FUEL                   89
COMPRESSED NATURAL GAS      27
Name: fuel_type, dtype: int64

---

### New Datasets

**Chicago Wards and Chicago census dataset**

In [6]:
# Chicago wards dataset
wards = pickle.load(
    open("D:/git_repositories/Datacamp-Joining_data_with_pandas/Datasets/ward.p", "rb")
)

In [21]:
# Print the shape of the wards
print("Total rows and columns of wards are: ", wards.shape)

Total rows and columns of wards are:  (50, 4)


In [22]:
wards.head()

ward            alderman                          address    zip
0    1  Proco "Joe" Moreno        2058 NORTH WESTERN AVENUE  60647
1    2       Brian Hopkins       1400 NORTH  ASHLAND AVENUE  60622
2    3          Pat Dowell          5046 SOUTH STATE STREET  60609
3    4    William D. Burns  435 EAST 35TH STREET, 1ST FLOOR  60616
4    5  Leslie A. Hairston            2325 EAST 71ST STREET  60649

In [7]:
# Chicago census dataset

census = pickle.load(
    open("D:/git_repositories/Datacamp-Joining_data_with_pandas/Datasets/census.p", "rb")
)

In [24]:
# Print the shape of the census
print("Total rows and columns of census dataset is: ", census.shape)

Total rows and columns of census dataset is:  (50, 6)


In [25]:
# Census Dataset
census.head()

ward  pop_2000  pop_2010 change                                  address  \
0    1     52951     56149     6%              2765 WEST SAINT MARY STREET   
1    2     54361     55805     3%                 WM WASTE MANAGEMENT 1500   
2    3     40385     53039    31%                      17 EAST 38TH STREET   
3    4     51953     54589     5%  31ST ST HARBOR BUILDING LAKEFRONT TRAIL   
4    5     55302     51455    -7%  JACKSON PARK LAGOON SOUTH CORNELL DRIVE   

     zip  
0  60647  
1  60622  
2  60653  
3  60653  
4  60637

#### Merging wards and census dataset

**Merge wards and census on the ward column and save the result to wards_census.**

In [8]:
# inner join
wards_census = wards.merge(
    census,
    on = 'ward',
    suffixes = ('_ward', '_cen')
)

In [27]:
# Print the shape of wards_census
print('wards_census table shape:', wards_census.shape)

wards_census table shape: (50, 9)


In [28]:
wards_census.head()

ward            alderman                     address_ward zip_ward  \
0    1  Proco "Joe" Moreno        2058 NORTH WESTERN AVENUE    60647   
1    2       Brian Hopkins       1400 NORTH  ASHLAND AVENUE    60622   
2    3          Pat Dowell          5046 SOUTH STATE STREET    60609   
3    4    William D. Burns  435 EAST 35TH STREET, 1ST FLOOR    60616   
4    5  Leslie A. Hairston            2325 EAST 71ST STREET    60649   

   pop_2000  pop_2010 change                              address_cen zip_cen  
0     52951     56149     6%              2765 WEST SAINT MARY STREET   60647  
1     54361     55805     3%                 WM WASTE MANAGEMENT 1500   60622  
2     40385     53039    31%                      17 EAST 38TH STREET   60653  
3     51953     54589     5%  31ST ST HARBOR BUILDING LAKEFRONT TRAIL   60653  
4     55302     51455    -7%  JACKSON PARK LAGOON SOUTH CORNELL DRIVE   60637

---

### Loading two new dataset

In [9]:
# Business owners dataset
biz_owners = pickle.load(
    open("D:/git_repositories/Datacamp-Joining_data_with_pandas/Datasets/business_owners.p", "rb")
)

In [32]:
biz_owners.head()

account first_name  last_name      title
0      10      PEARL    SHERMAN  PRESIDENT
1      10      PEARL    SHERMAN  SECRETARY
2   10002     WALTER     MROZEK    PARTNER
3   10002     CELINA     BYRDAK    PARTNER
4   10005      IRENE  ROSENFELD  PRESIDENT

In [10]:
# Licenses dataset
licenses = pickle.load(
    open("D:/git_repositories/Datacamp-Joining_data_with_pandas/Datasets/licenses.p", "rb")
)

In [36]:
licenses.head()

account ward  aid                   business               address    zip
0  307071    3  743       REGGIE'S BAR & GRILL       2105 S STATE ST  60616
1      10   10  829                 HONEYBEERS   13200 S HOUSTON AVE  60633
2   10002   14  775                CELINA DELI     5089 S ARCHER AVE  60632
3   10005   12  NaN  KRAFT FOODS NORTH AMERICA        2005 W 43RD ST  60609
4   10044   44  638  NEYBOUR'S TAVERN & GRILLE  3651 N SOUTHPORT AVE  60613

#### Merging these two datasets

In [11]:
# Merge the licenses and biz_owners table on account
licenses_owners = licenses.merge(
    biz_owners,
    on = 'account'
)

**Group licenses_owners by title and count the number of accounts for each title. Save the result as counted_df**

In [38]:
# grouping and counting
counted_df = licenses_owners.groupby('title').agg({'account':'count'})

In [40]:
# Sort the counted data in decending order
sorted_df = counted_df.sort_values('account', ascending = False)

In [42]:
# Use .head() method to print the first few rows of sorted_df
sorted_df.head()

account
title                   
PRESIDENT           6259
SECRETARY           5205
SOLE PROPRIETOR     1658
OTHER               1200
VICE PRESIDENT       970

---

### Loading 3 datasets

In [13]:
# CTA Ridership dataset
ridership = pickle.load(
    open("D:/git_repositories/Datacamp-Joining_data_with_pandas/Datasets/cta_ridership.p", "rb")
)

In [14]:
ridership.head()

station_id  year  month  day  rides
0      40010  2019      1    1    576
1      40010  2019      1    2   1457
2      40010  2019      1    3   1543
3      40010  2019      1    4   1621
4      40010  2019      1    5    719

In [15]:
# CTA calender dataset

cal = pickle.load(
    open("D:/git_repositories/Datacamp-Joining_data_with_pandas/Datasets/cta_calendar.p", "rb")
)

In [16]:
cal.head()

year  month  day        day_type
0  2019      1    1  Sunday/Holiday
1  2019      1    2         Weekday
2  2019      1    3         Weekday
3  2019      1    4         Weekday
4  2019      1    5        Saturday

In [17]:
# CTA Stations dataset

stations = pickle.load(
    open("D:/git_repositories/Datacamp-Joining_data_with_pandas/Datasets/stations.p", "rb")
)

In [18]:
stations.head()

station_id        station_name                 location
0      40010  Austin-Forest Park  (41.870851, -87.776812)
1      40020         Harlem-Lake  (41.886848, -87.803176)
2      40030        Pulaski-Lake  (41.885412, -87.725404)
3      40040        Quincy/Wells   (41.878723, -87.63374)
4      40050               Davis   (42.04771, -87.683543)

#### Merge the ridership and calender table

In [20]:
# Merging ridership and cal table

ridership_cal = ridership.merge(
    cal,
    on = ['year', 'month', 'day']
)

#### Merge ridership, calender and stations table together

In [21]:
ridership_cal_stations = ridership.merge(cal, on = ['year', 'month', 'day']) \
                            .merge(stations, on = 'station_id')

In [22]:
ridership_cal_stations.head()

station_id  year  month  day  rides        day_type        station_name  \
0      40010  2019      1    1    576  Sunday/Holiday  Austin-Forest Park   
1      40010  2019      1    2   1457         Weekday  Austin-Forest Park   
2      40010  2019      1    3   1543         Weekday  Austin-Forest Park   
3      40010  2019      1    4   1621         Weekday  Austin-Forest Park   
4      40010  2019      1    5    719        Saturday  Austin-Forest Park   

                  location  
0  (41.870851, -87.776812)  
1  (41.870851, -87.776812)  
2  (41.870851, -87.776812)  
3  (41.870851, -87.776812)  
4  (41.870851, -87.776812)

**Create a variable called filter_criteria to select the appropriate rows from the merged table so that you can sum the rides column.**

In [23]:
# Find out total number of rides provided to passengers passing through the Wilson station (station_name == 'Wilson') 
# when riding Chicago's public transportation system on weekdays (day_type == 'Weekday') in July (month == 7).
filter_criteria = (
    (ridership_cal_stations["station_name"] == 'Wilson') &
    (ridership_cal_stations["day_type"] == 'Weekday') &
    (ridership_cal_stations["month"] == 7)
)

In [25]:
total_number_rides = ridership_cal_stations.loc[filter_criteria, 'rides'].sum()
total_number_rides

140005

#### Three table merge

To solidify the concept of a three DataFrame merge, practice another exercise. A reasonable extension of our review of Chicago business data would include looking at demographics information about the neighborhoods where the businesses are. A table with the median income by zip code has been provided to you. You will merge the **licenses** and **wards** tables with this new **_income-by-zip-code_** table called **zip_demo**.

Starting with the licenses table, merge to it the zip_demo table on the zip column. Then merge the resulting table to the wards table on the ward column. Save result of the three merged tables to a variable named licenses_zip_ward

In [27]:
# Loading new dataset zip_demo

zip_demo = pickle.load(
    open("D:/git_repositories/Datacamp-Joining_data_with_pandas/Datasets/zip_demo.p", "rb")
)

In [28]:
zip_demo.head()

zip  income
0  60630   70122
1  60640   50488
2  60622   87143
3  60614  100116
4  60608   41226

In [31]:
licenses_zip_ward = licenses.merge(zip_demo, on = 'zip') \
                    .merge(wards, on = 'ward')

In [32]:
licenses_zip_ward.head()

account ward  aid               business                address_x  zip_x  \
0  307071    3  743   REGGIE'S BAR & GRILL          2105 S STATE ST  60616   
1   11280    3  763              PRIME WAY      2251 S STATE ST 1ST  60616   
2   15015    3  NaN  SOUTHVIEW MANOR, INC.      3311 S MICHIGAN AVE  60616   
3   19168    3  666               BP AMOCO  3101 S MICHIGAN AVE 1ST  60616   
4  205980    3  763  J & J  FISH & CHICKEN            8 E CERMAK RD  60616   

   income    alderman                address_y  zip_y  
0   46340  Pat Dowell  5046 SOUTH STATE STREET  60609  
1   46340  Pat Dowell  5046 SOUTH STATE STREET  60609  
2   46340  Pat Dowell  5046 SOUTH STATE STREET  60609  
3   46340  Pat Dowell  5046 SOUTH STATE STREET  60609  
4   46340  Pat Dowell  5046 SOUTH STATE STREET  60609

In [37]:
# find out the median income

licenses_zip_ward.groupby('alderman').agg( {'income' : 'median'} ).head()

income
alderman                       
Ameya Pawar             66246.0
Anthony A. Beale        38206.0
Anthony V. Napolitano   82226.0
Ariel E. Reyboras       41307.0
Brendan Reilly         110215.0

#### One to many merge with multiple tables

In [38]:
# Loading new dataset Chicago land use

land_use = pickle.load(
    open("D:/git_repositories/Datacamp-Joining_data_with_pandas/Datasets/land_use.p", "rb")
)

In [39]:
land_use.head()

ward  residential  commercial  industrial  vacant  other
0    1           41           9           2       2     46
1    2           31          11           6       2     50
2    3           20           5           3      13     59
3    4           22          13           0       7     58
4    5           25           3           1       3     68

In [40]:
census.head()

ward  pop_2000  pop_2010 change                                  address  \
0    1     52951     56149     6%              2765 WEST SAINT MARY STREET   
1    2     54361     55805     3%                 WM WASTE MANAGEMENT 1500   
2    3     40385     53039    31%                      17 EAST 38TH STREET   
3    4     51953     54589     5%  31ST ST HARBOR BUILDING LAKEFRONT TRAIL   
4    5     55302     51455    -7%  JACKSON PARK LAGOON SOUTH CORNELL DRIVE   

     zip  
0  60647  
1  60622  
2  60653  
3  60653  
4  60637

In [41]:
licenses.head()

account ward  aid                   business               address    zip
0  307071    3  743       REGGIE'S BAR & GRILL       2105 S STATE ST  60616
1      10   10  829                 HONEYBEERS   13200 S HOUSTON AVE  60633
2   10002   14  775                CELINA DELI     5089 S ARCHER AVE  60632
3   10005   12  NaN  KRAFT FOODS NORTH AMERICA        2005 W 43RD ST  60609
4   10044   44  638  NEYBOUR'S TAVERN & GRILLE  3651 N SOUTHPORT AVE  60613

**Merge land_use and census on the ward column. Merge the result of this with licenses on the ward column, using the suffix _cen for the left table and _lic for the right table. Save this to the variable land_cen_lic.**

In [42]:
# merging three tables

land_cen_lic = land_use.merge(census, on = 'ward') \
                .merge(licenses, on = 'ward', suffixes = ['_cen', '_lic'])

In [43]:
land_cen_lic.head()

ward  residential  commercial  industrial  vacant  other  pop_2000  \
0    1           41           9           2       2     46     52951   
1    1           41           9           2       2     46     52951   
2    1           41           9           2       2     46     52951   
3    1           41           9           2       2     46     52951   
4    1           41           9           2       2     46     52951   

   pop_2010 change                  address_cen zip_cen account  aid  \
0     56149     6%  2765 WEST SAINT MARY STREET   60647   12024  NaN   
1     56149     6%  2765 WEST SAINT MARY STREET   60647   14446  743   
2     56149     6%  2765 WEST SAINT MARY STREET   60647   14624  775   
3     56149     6%  2765 WEST SAINT MARY STREET   60647   14987  NaN   
4     56149     6%  2765 WEST SAINT MARY STREET   60647   15642  814   

               business              address_lic zip_lic  
0   DIGILOG ELECTRONICS       1038 N ASHLAND AVE   60622  
1      EMPTY BOTTLE INC   1035 N WESTERN AVE 1ST   60622  
2  LITTLE MEL'S HOT DOG    2205 N CALIFORNIA AVE   60647  
3    MR. BROWN'S LOUNGE   2301 W CHICAGO AVE 1ST   60622  
4          Beat Kitchen  2000-2100 W DIVISION ST   60622

**Group land_cen_lic by ward, pop_2010 (the population in 2010), and vacant, then count the number of accounts. Save the results to pop_vac_lic.**

In [51]:
pop_vac_lic = land_cen_lic.groupby(['ward', 'pop_2010', 'vacant'], as_index = False).agg({'account':'count'})

In [52]:
pop_vac_lic.head()

ward  pop_2010  vacant  account
0    1     56149       2      253
1   10     51535      14      130
2   11     51497       5      201
3   12     52235       4      255
4   13     53722       1      101

Sort **pop_vac_lic** by vacant, account, andpop_2010 in descending, ascending, and ascending order respectively. Save it as **sorted_pop_vac_lic.**

In [53]:
sorted_pop_vac_lic = pop_vac_lic.sort_values(['vacant', 'account', 'pop_2010'], ascending = [False, True, True])
sorted_pop_vac_lic.head()

ward  pop_2010  vacant  account
47    7     51581      19       80
12   20     52372      15      123
1    10     51535      14      130
16   24     54909      13       98
7    16     51954      13      156

---